### ISSUE TITLE
Clipped input images

### ISSUE THREAD

**Issue by vitruvi** [ustc-bmec/fMRI-Conv-Att#1]
Can you tell how you clip the (91x109x91) volume and obtain the (80, 96, 88) volume?

**Comment by ustc-bmec** [comment]
The blank parts that did not contain brain tissues were cropped out and the (80, 96, 88) volume was obtained.

**Comment by vitruvi** [comment]
Can you explicitly define how many rows and columns you discard on each end of the volume border? As in "Whole-brain-conv", data = data[8:-8, 8:-8, :-10, frames].

I have tried an approximate clipping of borders, but the result is very low accuracy. To reproduce this work, knowing the exact number of clipped border lines/rows is crucial. I am running a code to find the empty border lines on a limited set of samples and what I find might be different than using all 1200 subjects.

I had also reproduced the previous project, Whole-brain-conv. But there is also a reproducibility problem on that project when you run the demo jupyter notebook. I couldn't find the source of the problem.

**Comment by ustc-bmec** [comment]
The data used is the HCP minimum preprocessed. So the variation of the outline of the brain exists but is very small. 
The advatange of CNN is that it is insensitive to the edge croping. So the issue may be related to the compatibility.
I'm sure that the code worked before it was uploaded to Github.
I'll try to find the problem in the coming days.

**Comment by vitruvi** [comment]
Can you also produce a requirements.txt via "pip freeze > requirements.txt" in the working environment?

How does the model converge in your case? I couldn't get an improvement in validation set after 4 epochs.

**Comment by xiaoshi0107** [comment]
> Can you explicitly define how many rows and columns you discard on each end of the volume border? As in "Whole-brain-conv", data = data[8:-8, 8:-8, :-10, frames].
> 
> I have tried an approximate clipping of borders, but the result is very low accuracy. To reproduce this work, knowing the exact number of clipped border lines/rows is crucial. I am running a code to find the empty border lines on a limited set of samples and what I find might be different than using all 1200 subjects.
> 
> I had also reproduced the previous project, Whole-brain-conv. But there is also a reproducibility problem on that project when you run the demo jupyter notebook. I couldn't find the source of the problem.

How many data do you use when you reproduce the network. The amount of data is very important.

**Comment by vitruvi** [comment]
> > Can you explicitly define how many rows and columns you discard on each end of the volume border? As in "Whole-brain-conv", data = data[8:-8, 8:-8, :-10, frames].
> > I have tried an approximate clipping of borders, but the result is very low accuracy. To reproduce this work, knowing the exact number of clipped border lines/rows is crucial. I am running a code to find the empty border lines on a limited set of samples and what I find might be different than using all 1200 subjects.
> > I had also reproduced the previous project, Whole-brain-conv. But there is also a reproducibility problem on that project when you run the demo jupyter notebook. I couldn't find the source of the problem.
> 
> How many data do you use when you reproduce the network. The amount of data is very important.

I just want to see the performance of the saved checkpoint model on a subset of the HCP dataset, not for training. (Since the approximate clipping didn't work with the saved checkpoint, I tried to train it from scratch. But it's not my main goal.) So I need to know how to clip the HCP data, so that the model in this repository can accurately classify them.

I am checking if there is a compatibility problem or a problem with model convergence.

**Comment by vitruvi** [comment]
> The data used is the HCP minimum preprocessed. So the variation of the outline of the brain exists but is very small. The advatange of CNN is that it is insensitive to the edge croping. So the issue may be related to the compatibility. I'm sure that the code worked before it was uploaded to Github. I'll try to find the problem in the coming days.

If I knew how many lines are clipped from each border, it would speed things up for me. Translation invariance may not be totally applicable in some cases.

**Comment by vitruvi** [comment]
I have completed the empty mask run on HCP and the non-empty borders are data[8:-9,8:-7,:-9] of size (82,94,74). Could it be that (80, 96, 88) is the common mask size of BOLD5000 and HCP?

**Comment by vitruvi** [comment]
I generated the data samples from minimally preprocessed HCP files using the script in Whole_Brain_Conv project. 

I get the following confusion matrix, after loading the provided checkpoint file and evaluating each sample with 27 time-points per sample.

Accuracy: 0.5539192113488819

[[1619  264    1    8   11   58   22]
 [ 104  836    0    5    1   24    5]
 [ 520  413  700   42    6  144  129]
 [ 289  331    7 1524   43  121   80]
 [ 929  795    4   36  935  111   34]
 [ 834  327    4   37   16 1413   44]
 [ 634  580    5  164   30  209 2188]]

I would appreciate so much if you could kindly help me reproduce the journal's accuracy values.

**Comment by ustc-bmec** [comment]
The cropping is [6:-5, 7:-6, :-6, frames] or [5:-6, 6:-7, :-6, frames], both are fine.

**Comment by vitruvi** [comment]
That makes 80:96:85 from 91:109:91. Journal reports 80:96:88. I'm assuming you mean 5:-6, 6:-7, :-3. Is that right?

Can you simply run the checkpoint model on any approx. 800 HCP subject data (no matter inside or outside training set) and tell the accuracy or confusion matrix? In my case, it is around 55pct.

I want to find out what I should change. Data or software environment.

pipreqs --force or pip freeze to generate "requirements.txt" would also help a lot. This file lists the packages installed in the original environment that produced the high accuracy result and, with it, I can also reproduce the same environment to check the source of the reproducibility error.

**Comment by ustc-bmec** [comment]
Happy Chinese new year!
Sorry, I made a mistake and the -3 is right.
It's during the Chinese new year holidays. So it is not easy to contact with the guy writting the codes. I will response in a few days, by the end of the holiday.

**Comment by vitruvi** [comment]
The reproducibility problem still persists and a reviewer asks me to reproduce your results. Why do you think the pretrained model is not working as reported in the article?

**Comment by wuyueyang817** [comment]
Make sure that the data you use matches the data described in the paper, such as 
the HCP S1200 minimum preprocessed dataset that has been normalized to MNI 152 space, 
the time period of the clipped data contains an hrf response, rather than being randomly selected or including all rest periods;
confirm that the condition of each task you clipped aligns with the description provided in the paper.
and notice that frames used here is 15 discribed in dataset.py not 27.
and the amount of data clips you use is more than 14,821 as discribed, it,s a good thought to check your dataset again.

**Comment by vitruvi** [comment]
Do you follow the same routine as Wang et. al. (whole-brain-conv in your repository) to split a session into subtasks? Are there any differences?

Quoting from the journal: "A segment of k continuous frames, which was randomly split from each instance, was used as input for training. During the testing stages, the predictions for all segmentations of one instance are summed up, and the task label with the majority vote is predicted to represent the final class of the instance."

Is the "instance" in the quote a subtask duration? For example, fear subtask of emotion task has 37 frames. You train on random 17 consecutive frames during training. Then, you evaluate every consecutive 17 frame subsets of the 37 frames and assign the majority label vote to all 37 frames. (You evaluate all t in X[t-17, t], where t is in the interval [18, 37]) Is this right?

**Comment by wuyueyang817** [comment]
yeah, what you said is right.

**Comment by vitruvi** [comment]
Thanks for your answer to my second question.

Can you also answer the first question? Your answer will clear the details of the data processing differences, if any.

Do you follow the same routine as Wang et. al. (whole-brain-conv in your repository) to split a session into subtasks? Are there any differences?

**Comment by wuyueyang817** [comment]
I haven't follow the rotine, but read the paper. 
In whole-brain-conv, "For each task, an input sample was a continuous BOLD series that covered the entire block and 8 s past the block, including the post-signal of the hemodynamic response function (HRF). Furthermore, each BOLD volume was cropped from 91 × 109 × 91 to 75 × 93 × 81 to exclude the area that was not part of the brain. Thus, the input data varied from 27 × 75 × 93 × 81 to 50 × 75 × 93 × 81 (time × x × y × z, TR = 0.72 s). A total of 34,938 fMRI 4D data items were obtained across all tasks and subjects." while "only the fragment consisting of the first k TRs of each data was used in validation and testing stages." where k is 27.
but here, "A segment of k continuous frames, which was randomly split from each instance, was used as input for training. During the testing stages, the predictions for all segmentations of one instance are summed up, and the task label with the majority vote is predicted to represent the final class of the instance".
I haven't try on my own, and dont know your pipeline, but I guess there are 2 probable reasons:
1. The data here probably doesn't include 8s past the block;
2. Only the previous 27 frame is used in validation stage of whole-brain-conv while here is "voting", you know the signal of past 8s of the hrf block may be of lower SNR, and 8s may be too long for 15 frames in this paper.

**Comment by vitruvi** [comment]
Following the setup of Wang et. al. (whole brain conv project in the repository), I just split each session into smaller clips such that each clip includes data of only one subtask (i.e. only rh for motor task), using AFNI 3dTcat. There is no additional step in my data preparation.

Could you kindly check the model on last 15 frames of each subtask duration?
Below is the evaluation result for 1000/35010 samples (only first class is predicted by the model). 

at sample: 1001/35016
                    precision    recall  f1-score   support
                 0       1.00      0.15      0.26      1001
                 1       0.00      0.00      0.00         0
                 2       0.00      0.00      0.00         0
                 4       0.00      0.00      0.00         0
                 5       0.00      0.00      0.00         0
                 6       0.00      0.00      0.00         0
    accuracy                    0.15      1001
   macro avg     0.17      0.02      0.04      1001
weighted avg   1.00      0.15      0.26      1001

[[147 576  54   1  48 175]
 [  0   0   0   0   0   0]
 [  0   0   0   0   0   0]
 [  0   0   0   0   0   0]
 [  0   0   0   0   0   0]
 [  0   0   0   0   0   0]]


Another thing is the difference in provided samples and HCP minimally preprocessed data. The sample data on the samples folder is below. Since I don't know the session of the sample, I compare it with both LR and RL sessions.
Even if the data are the same (it is not possible to tell just by observation), the brain masks are different in the images below.

This is the FSLEyes output of the provided sample (100206_MOTOR_rh.nii.gz) in github project.
![100206_MOTOR_rh nii gz](https://github.com/ustc-bmec/fMRI-Conv-Att/assets/59297198/0e196cf2-4807-4b09-bcaa-a8b7e1f4aa90)

Below are the images for LR and RL session MOTOR rh subtask of subject 100206.

![100206_MOTOR_rh nii gz_RL_local](https://github.com/ustc-bmec/fMRI-Conv-Att/assets/59297198/3778f0c5-ccb3-4c5b-a9f3-7b6d5ac3f60e)

![100206_MOTOR_rh nii gz_LR_local](https://github.com/ustc-bmec/fMRI-Conv-Att/assets/59297198/8dae0f54-82c4-4d93-98f6-90a51e27f502)

Below is the basic python script to evaluate the model on HCP samples.

```python
import os
import sys
import torch
import numpy as np
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
import pandas
from tqdm import tqdm

from model import Attention_network
from dataset import BrainDataset

device = torch.device('cpu' if not torch.cuda.is_available() else 'cuda')

df = pandas.read_csv(sys.argv[1] + '/index.txt', names=["subj", "task", "block", "index", "path"])

# parse subtask name, create new column
df["sub"] = [d.split("_")[-1][:-7] for d in df["path"]]

# list of tasks and subjects overall
tasks = df["task"].unique()
subjects = df["subj"].unique()

# unique subjects per task
subjs = []
for idx, task in enumerate(tasks):
    subjs.extend([df.loc[df["task"] == tasks[idx]]["subj"].unique()])

# set of common subjects for all tasks(some subjects have faulty images of one task, that is removed. omitting such subjects.)
common = set.intersection(set(df.loc[df["task"] == tasks[0]]["subj"].unique()),
                          set(df.loc[df["task"] == tasks[1]]["subj"].unique()),
                          set(df.loc[df["task"] == tasks[2]]["subj"].unique()),
                          set(df.loc[df["task"] == tasks[3]]["subj"].unique()),
                          set(df.loc[df["task"] == tasks[4]]["subj"].unique()),
                          set(df.loc[df["task"] == tasks[5]]["subj"].unique()),
                          set(df.loc[df["task"] == tasks[6]]["subj"].unique()))

tmp = df.loc[df["subj"].isin(common)]
train_files = np.array(tmp["path"])
train_lbls = tmp["task"]

onehot = np.eye(7)
train_lbls = np.array([onehot[np.where(tasks == t)] for t in train_lbls])

dataset = BrainDataset(train_files, train_lbls, is_train=False)

model = Attention_network().to("cuda")
model.load_state_dict(torch.load('./pre_train.pth.tar')['state_dict'])
_ = model.eval()

preds = []
reals = []

with torch.no_grad():
    for idx, data in enumerate(tqdm(dataset)):
        # data = dataset[0]
        inputs, labels = torch.tensor(data[0], dtype=torch.float).to(device), \
            torch.tensor(data[1], dtype=torch.long).squeeze().to(device)
        inputs = torch.unsqueeze(inputs,0)
        pred_prob = model(inputs)
        prelable = torch.max(pred_prob, dim=1)[-1].cpu().numpy()

        preds.extend(np.expand_dims(np.argmax(pred_prob.detach().cpu(), axis=1), axis=-1))
        reals.extend(np.expand_dims(labels.detach().cpu(), axis=-1))

        if idx % 1000 == 0:
            print("\nat sample: %d" % (idx))
            cr = classification_report(reals, preds)
            print(cr)
            cm = confusion_matrix(reals, preds)
            print(cm)
np.save("preds.npy", preds)
np.save("reals.npy", reals)
```

(I made a mistake in the code that led me to think the pretrained model worked, so I had to change the comment.)

**Comment by vitruvi** [comment]
I made a mistake in the code that led me to think the pretrained model worked. Reopening the issue.
